pkg/controller/pravegacluster/upgrade.go

```diff
@@ -404,6 +405,178 @@ func (r *ReconcilePravegaCluster) syncSegmentStoreVersion(p *pravegav1alpha1.Pra
+		if err != nil {
+			// Abort if there is any errors with the updated pods
+			return false, fmt.Errorf("updating statefulset (%s) failed due to %v", newsts.Name, err)
+		}
+	}
+
+	//this check to ensure that the oldsts always decrease by 2 as well as newsts pods increase by 2 only then the next increment or decrement happen
+	//first condition is true and used in the case of rollbackCondition
+	//second condition is true and used in case of Upgrade
+	if r.rollbackConditionFor07(p, newsts) || r.upgradeConditionFor07(p, newsts, oldsts) {
+		//this check is run till the value of old sts replicas is greater than 0 and will increase two replicas of the new sts and delete 2 replicas of the old sts
+		if *oldsts.Spec.Replicas > 2 {
+			err = r.incrementWhenReplicasMoreThan2(p, newsts, oldsts)
+			if err != nil {
+				return false, err
+			}
+		} else {
+			//here we remove the pvc's attached with the old sts and deleted it when old sts replicas have become 0
+			err = r.incrementWhenReplicasLessThan2(p, newsts, oldsts)
+			if err != nil {
+				return false, err
+			}
+		}
+	}
+	//upgrade is still in process
+	return false, nil
+}
+
+//this function will check if furter increment or decrement in pods needed in case of rollback from version0.7
+func (r *ReconcilePravegaCluster) rollbackConditionFor07(p *pravegav1alpha1.PravegaCluster, sts *appsv1.StatefulSet) bool {
+	if r.IsClusterRollbackingFrom07(p) && sts.Status.ReadyReplicas == *sts.Spec.Replicas {
+		return true
+	}
+	return false
+}
+
+//this function will check if furter increment or decrement in pods needed in case of upgrade to version0.7 or above
+func (r *ReconcilePravegaCluster) upgradeConditionFor07(p *pravegav1alpha1.PravegaCluster, newsts *appsv1.StatefulSet, oldsts *appsv1.StatefulSet) bool {
+	if oldsts.Status.ReadyReplicas+newsts.Status.ReadyReplicas == p.Spec.Pravega.SegmentStoreReplicas && newsts.Status.ReadyReplicas == *newsts.Spec.Replicas {
+		return true
+	}
+	return false
+}
+
+//this function will increase two replicas of the new sts and delete 2 replicas of the old sts everytime it's called
+func (r *ReconcilePravegaCluster) incrementWhenReplicasMoreThan2(p *pravegav1alpha1.PravegaCluster, newsts *appsv1.StatefulSet, oldsts *appsv1.StatefulSet) error {
+	*newsts.Spec.Replicas = *newsts.Spec.Replicas + 2
+	err := r.client.Update(context.TODO(), newsts)
+	if err != nil {
+		return fmt.Errorf("updating statefulset (%s) failed due to %v", newsts.Name, err)
+	}
+	*oldsts.Spec.Replicas = *oldsts.Spec.Replicas - 2
+	err = r.client.Update(context.TODO(), oldsts)
+	if err != nil {
+		return fmt.Errorf("updating statefulset (%s) failed due to %v", oldsts.Name, err)
+	}
+	return nil
+}
+
+//This function will remove the pvc's attached with the old sts and deleted it when old sts replicas have become 0
+func (r *ReconcilePravegaCluster) incrementWhenReplicasLessThan2(p *pravegav1alpha1.PravegaCluster, newsts *appsv1.StatefulSet, oldsts *appsv1.StatefulSet) error {
```

Comment:
Please use a better method name that is indicative of what the method does..... Perhaps transitionToNewSTS() ???

Reply:
Changed it